### DISCUSSION TITLE
Loop Inside class

### DISCUSSION THREAD

**Discussion by Nagateja23** [mesa/mesa#1373] [Q&A]
So I have a csv file with some parameters, I want to use those parameters and run my model again and again and store the results of the parameters in a csv.
My current code is
`class Opiodmodel(model):`
    `def __init__(self, num_agents=40, width=5, height=5):`
     .
     .
     `def step(self):`
     .
     .
     .

So instead of fixing num_agents,width,height I want to use my csv and take it from there. It has multiple num_agents, widths and heights I want to loop over all of them and store results of all of them...
I tried doing this:

`class Opiodmodel(model):`
    `for i in range(len(csvfile)):`
        `n = csvfile['n'][i]`
        `w = csvfile['width'][i]`
        `h = csvfile['height'][i]`
       `def __init__(self, num_agents=n, width=w, height=h):`
     .
     .
        `def step(self):`
     .
     .
     .
But for some reason, the model is only taking in the last row of csv file, it's not iterating over all the values.
Any suggestions are appreciated.

**Comment by rht** [comment]
You need to describe how you read the CSV file. That said, you are probably reading the CSV file wrong. Just use one of the method in https://www.geeksforgeeks.org/read-a-csv-into-list-of-lists-in-python/ -- as in, read the whole file into a list of list. Or use Pandas (also described in the URL I linked).

**Reply by rht** [reply]
Also, you are not supposed to loop the parameters inside the class definition. You need to do it while creating a new model object every time.

**Reply by Nagateja23** [reply]
Sorry for not describing how I read the csv file
```
for i in range(len(sample)):
    n = sample['n'][i]
    w = sample['width'][i]
    h = sample['height'][i]
    stop = sample['end_time'][i]
    parameter_set = sample['parameter_set'][i]
    grid = CanvasGrid(Opioid_draw, h, w, 500, 500)
    model_params = {
        "num_agents":n,
        "height": h,
        "width": w,
    }

    server = ModularServer(OpioidModel, [grid,abm_chart], "Opioid",model_params)
```
I tried to do this, but as expected, when server is launched it's only launched for the last entry of the csv.
Then I tried to create a list and append all the Modular instances in it and tried assigning a different port to them ```server[i].port() = 8521+i```, but I cannot assign a function call to ```server[i]``` as well.
My end goal is to create a server for all the entries of the csv file, once I stop the model for the first entry it closes the server and server for the next entry is open.
I'll be very grateful of you give me a suggestion or point me in the direction of how I can achieve this goal.

**Reply by rht** [reply]
You can assign different ports by passing in the port parameter you want to the `server.launch()` method, but this means you have N servers simultaneously running.

If you want to launch them in order, you have to make it so that you don't immediately create the next server in your loop. I think this may suffice, but haven't tested:
```python
for i in range(len(sample)):
    ...
    server = ModularServer(....)
    server.launch(theportyouwant)
```
The `server.launch()` line should be blocking. Once you press Ctrl-C to stop the current one, it should go to the next one.

**Reply by Nagateja23** [reply]
Thanks for the quick response, but I have already tried this for some reason ```Ctrl+C``` is not stopping the server and if i click ```x``` and stop the server manually, the code stops it doesn't launch another server instance.
```
for i in range(len(sample)):
    n = sample['n'][i]
    w = sample['width'][i]
    h = sample['height'][i]
    stop = sample['end_time'][i]
    parameter_set = sample['parameter_set'][i]
    grid = CanvasGrid(Opioid_draw, h, w, 500, 500)
    model_params = {
        "num_agents":n,
        "height": h,
        "width": w,
    }

    server = ModularServer(OpioidModel, [grid,abm_chart], "Opioid",model_params)
    server.launch(8521+i)
```

**Reply by rht** [reply]
I suppose that's because Ctrl-C kill the entire Python process, instead of just the modular server. This is the relevant code for server.launch:
https://github.com/projectmesa/mesa/blob/60bc768f5eede2dc4806684ae4b0b3046eb07d27/mesa/visualization/ModularVisualization.py#L356-L366
You have to somehow catch a keyboard event, and then you have trigger https://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop.stop. Or alternatively, add a web UI button that lets you stop the server. IMO I don't think this is worth the effort.

I think you can just store the states of the grid throughout the simulation, for each parameter, and you can animate them with Matplotlib later instead.
